### PR TITLE
Efficiency fixes

### PR DESCRIPTION
**Pull request by Average-Bear**
I've made some general changes. Please, review and TEST prior to accepting changes.

- Changed the $rpt variable to $FinalReport, then removed the += operator and replaced them with .Add()
- Removed multiple uses of Get-ADUser and replaced it with the already existing array of $AllUsers and filtered requested values locally
- Added synopsis, description, and parameter information
- General changes to formatting

### CODE REVIEW

File: PSHTML-AD.ps1 [hunk 1]
```diff
@@ -125,253 +185,256 @@ $EnterpriseAdminTable 				= New-Object 'System.Collections.Generic.List[System.O
...
+$TOPUserTable                       = New-Object 'System.Collections.Generic.List[System.Object]'
+$TOPGroupsTable                     = New-Object 'System.Collections.Generic.List[System.Object]'
+$TOPComputersTable                  = New-Object 'System.Collections.Generic.List[System.Object]'
+$GraphComputerOS                    = New-Object 'System.Collections.Generic.List[System.Object]'
+
+#Get all users right away. Instead of doing several lookups, we will use this object to look up all the information needed.
 $AllUsers = Get-ADUser -Filter * -Properties *
-$GPOs = Get-GPO -all | Select-Object DisplayName, GPOStatus, ModificationTime, @{ Label = "ComputerVersion"; Expression = { $_.computer.dsversion } }, @{ Label = "UserVersion"; Expression = { $_.user.dsversion } }
+
+$GPOs = Get-GPO -aAl | Select-Object DisplayName, GPOStatus, ModificationTime, @{ Label = "ComputerVersion"; Expression = { $_.computer.dsversion } }, @{ Label = "UserVersion"; Expression = { $_.user.dsversion } }
```
Comment (furicle) [inline PSHTML-AD.ps1]: TYPO!  -all not -aAl
```suggestion
$GPOs = Get-GPO -all | Select-Object DisplayName, GPOStatus, ModificationTime, @{ Label = "ComputerVersion"; Expression = { $_.computer.dsversion } }, @{ Label = "UserVersion"; Expression = { $_.user.dsversion } }
```

**Comment by furicle** [comment]
Not my project, not my patch, so feel free to ignore, but

I think there are three or four different patches here rather than one big one.

Adding parameters is one
Changing all the foreach styles is another
Changing += to .Add( is a third
The Get-ADUser stuff is a fourth

If they are separate, they're easier to review, debug and accept.

And some of it is style, and some of it is functional - the original author might want to pick and choose what he thinks matches his project best.

Just my 0.02 cents - hopefully taken as constructive as I'd hate to dissuade anyone from contributing!

**Comment by bwya77** [comment]
> Not my project, not my patch, so feel free to ignore, but
> 
> I think there are three or four different patches here rather than one big one.
> 
> Adding parameters is one
> Changing all the foreach styles is another
> Changing += to .Add( is a third
> The Get-ADUser stuff is a fourth
> 
> If they are separate, they're easier to review, debug and accept.
> 
> And some of it is style, and some of it is functional - the original author might want to pick and choose what he thinks matches his project best.
> 
> Just my 0.02 cents - hopefully taken as constructive as I'd hate to dissuade anyone from contributing!

are you wanting foreach ($item in $Items) to be changed to foreach-object?
In merge not finding +=
Fixed get-aduser as well as get-gpo doing it multiple times

**Comment by Average-Bear** [comment]
I didn't think about submitting individual changes, I went through one mass sweep and made the changes on the local version of the original I pulled down -- my apologies.

So, foreach() is likely better in this case. Foreach-Object is slightly different but accomplishes the same type of thing but, both have their place.

Think of foreach() and For-EachObject in the context of retrieving and watching a movie. You can do one of two things, respectively, download the entire movie and then watch it or stream it and watch it while it continues to pull down data. You would be doing double work however if you downloaded the entire thing, then started streaming the information -- if you already have it downloaded, just watch the downloaded version. 

Does this make sense? To be more clear, foreach() is downloading then watching and For-EachObject is streaming.

**Pull request merged**